From bda9bc7d28574ce8312965adcf530caa4405ae90 Mon Sep 17 00:00:00 2001
From: Damien Miller <djm@mindrot.org>
Date: Thu, 15 Oct 2026 16:56:13 +1100
Subject: [PATCH] readpassphrase.c fix from OpenBSD

revision 1.31
date: 2026/10/15 05:55:06;  author: djm;  state: Exp;  lines: +5 -5;  commitid: EQzOxkBlPs867NeQ;
fix weird behaviour after SIGINT/SIGQUIT caused by my previous
signal handling change. Feedback deraadt/millert; ok deraadt

(cherry picked from commit bda9bc7d28574ce8312965adcf530caa4405ae90)
---
 openbsd-compat/readpassphrase.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/openbsd-compat/readpassphrase.c b/openbsd-compat/readpassphrase.c
index 43b7128..4b710b4 100644
--- a/openbsd-compat/readpassphrase.c
+++ b/openbsd-compat/readpassphrase.c
@@ -1,4 +1,4 @@
-/*	$OpenBSD: readpassphrase.c,v 1.30 2026/09/16 01:24:59 djm Exp $	*/
+/*	$OpenBSD: readpassphrase.c,v 1.31 2026/10/15 05:55:06 djm Exp $	*/
 
 /*
  * Copyright (c) 2000-2002, 2007, 2010
@@ -133,11 +133,11 @@ restart:
 	sigemptyset(&sa.sa_mask);
 	sa.sa_flags = 0;		/* don't restart system calls */
 	sa.sa_handler = handler;
+	sigaction(SIGINT, &sa, &saveint);
+	sigaction(SIGQUIT, &sa, &savequit);
 	sigaction_except_ign(SIGALRM, &sa, &savealrm);
 	sigaction_except_ign(SIGHUP, &sa, &savehup);
-	sigaction_except_ign(SIGINT, &sa, &saveint);
 	sigaction_except_ign(SIGPIPE, &sa, &savepipe);
-	sigaction_except_ign(SIGQUIT, &sa, &savequit);
 	sigaction_except_ign(SIGTERM, &sa, &saveterm);
 	sigaction_except_ign(SIGTSTP, &sa, &savetstp);
 	sigaction_except_ign(SIGTTIN, &sa, &savettin);
@@ -175,10 +175,10 @@ restart:
 			continue;
 		signo[SIGTTOU] = sigttou;
 	}
+	sigaction(SIGINT, &saveint, NULL);
+	sigaction(SIGQUIT, &savequit, NULL);
 	sigaction_except_ign(SIGALRM, &savealrm, NULL);
 	sigaction_except_ign(SIGHUP, &savehup, NULL);
-	sigaction_except_ign(SIGINT, &saveint, NULL);
-	sigaction_except_ign(SIGQUIT, &savequit, NULL);
 	sigaction_except_ign(SIGPIPE, &savepipe, NULL);
 	sigaction_except_ign(SIGTERM, &saveterm, NULL);
 	sigaction_except_ign(SIGTSTP, &savetstp, NULL);
-- 
2.53.0

